RD-754 Added labels widget

kubama opened the pull request:
NOTE: Table is not paginated as corresponding manager API does not support it.

TODOs:
- Add documentation
- Add thumbnail image
- Add system test
- Handle Cloudify prefixes

Widget in read only view (no `deployment_create` permission):
![image](https://user-images.githubusercontent.com/49753069/108824938-a6f34c80-75c2-11eb-9109-548985c9dd96.png)

Widget in standard view:
![image](https://user-images.githubusercontent.com/49753069/108826647-d7d48100-75c4-11eb-85ca-230aa0eba725.png)

Inline edit mode:
![image](https://user-images.githubusercontent.com/49753069/108826747-fc305d80-75c4-11eb-87a0-e2db4bae2647.png)

Add modal (note the modified header and empty labels list):
![image](https://user-images.githubusercontent.com/49753069/108827600-1e76ab00-75c6-11eb-8e77-3ebdf20b2058.png)

Gelio reviewed (commented):
1. How about using TypeScript for the new files you add :slightly_smiling_face: ?
2. Are you planning on filling in the README?
3. A video showing the workflow would be great, as it shows far more than just screenshots :smile: 

How can I test this widget in the UI? I changed the permissions and displayed the widgets, but it looks as follows:
![image](https://user-images.githubusercontent.com/889383/108850489-b8991c00-75e3-11eb-9475-43f41862df1d.png)
and I did not go deep enough to go further :slightly_smiling_face: 

Good job overall, I like that there is component reuse and most of the code is easy to follow :slightly_smiling_face:

qooban commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -73,11 +84,18 @@ function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
 
 ManageLabelsModal.propTypes = {
     deploymentId: PropTypes.string.isRequired,
+    existingLabels: PropTypes.arrayOf(),
```
That should probably be extended to:
```suggestion
    existingLabels: PropTypes.arrayOf(PropTypes.string),
```
?

kubama replied:
I reused LabelsPropType here.

qooban commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -53,7 +59,12 @@ function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
                     >
-                        <LabelsInput initialLabels={initialLabels} onChange={onChange} toolbox={toolbox} />
+                        <LabelsInput
+                            addMode={!!existingLabels}
```
In case of arrays/objects, I prefer `!_.isEmpty`, but I'm leaving it to you.

Gelio replied:
:thinking: but those 2 are different

```ts
const obj = {};
console.log(!!obj, !_.isEmpty(obj)); // true, false
```

kubama replied:
It wouldn't be semantically equal.

qooban commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -38,13 +40,17 @@ function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
         setLoading();
 
-        actions.doSetLabels(deploymentId, labels).then(onHide).catch(setMessageAsError).finally(unsetLoading);
+        actions
```
Just FYI - we'll have conflict here. I had to change it a bit in https://github.com/cloudify-cosmo/cloudify-stage/pull/1209 . Should be easy to resolve.

qooban commented on widgets/labels/widget.css:
```diff
@@ -0,0 +1,9 @@
+}
+
+.labelsWidget .labelsTable {
+    table-layout: fixed;
```
Why we need this one? 
Related to inline editing?

kubama replied:
This enabled me to set fixed width for actions column. It has only two icons, so there is no need to expand more than being able to hold these icons in line. On the other hand there are only two other columns, so there is plenty space in this table - no need for actions column to shrink (it would line up icons vertically). Summing up I decided fixed width is ideal for actions column.

qooban commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                toolbox={toolbox}
+            />
+
+            <Confirm
+                open={!!labelToDelete}
+                onCancel={unsetLabelToDelete}
+                onConfirm={() => {
+                    actions
+                        .doSetLabels(
+                            data.deploymentId,
+                            data.items.filter(filteredItem => filteredItem !== labelToDelete)
+                        )
+                        .then(() => toolbox.refresh());
+                    unsetLabelToDelete();
+                }}
+                content={
+                    <div className="content">
+                        Are you sure you want to remove label &nbsp;
+                        <Label size="large">
+                            {labelToDelete?.key} <span style={{ fontWeight: 'lighter' }}>{labelToDelete?.value}</span>
+                        </Label>
+                        ?
+                    </div>
+                }
+            />
```
1. I would extract it to a separate component.
2. You can move strings to `en.json` file at once.

kubama replied:
> 1. I would extract it to a separate component.

Done.

> 2. You can move strings to `en.json` file at once.

As far as I remember we agreed that en.json will contain strings for common components, shared across widgets (`widgets/common/src/**/*`) while strings specific to particular widgets should be kept with the widgets themselves (i18n support yet to be added).

qooban replied:
True. So for now (this PR) let's keep it as it is.

And let's discuss on UI weekly if it would be worth to change anything in that approach and maybe add support for i18n to new widgets only.

kubama commented:
> 1. How about using TypeScript for the new files you add 🙂 ?

I just don't feel brave enough 🙂.

> 2. Are you planning on filling in the README?

Yes, it's on the TODO list in PR description.

> 3. A video showing the workflow would be great, as it shows far more than just screenshots 😄

I'll try to keep it in mind for the next time.

> How can I test this widget in the UI? I changed the permissions and displayed the widgets, but it looks as follows:
> ![image](https://user-images.githubusercontent.com/889383/108850489-b8991c00-75e3-11eb-9475-43f41862df1d.png)
> and I did not go deep enough to go further 🙂

It works in deployment context, so the easiest way is to go to deployment and then add it to the page.

qooban commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                content={
+                    <div className="content">
+                        Are you sure you want to remove label &nbsp;
+                        <Label size="large">
```
This is the same approach as in the popup shown on hovering long label in `LabelsInput`, right?
If so, the we can probably have it externalized as a component and use it here and in `LabelsInput`.

kubama replied:
Content of the popup you are referring to is `<strong>{key}</strong> {value}`, so it's not the same. Even if it was, I think it's so simple it's not really worth externalizing.

qooban commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                            data.deploymentId,
+                            data.items.filter(filteredItem => filteredItem !== labelToDelete)
+                        )
+                        .then(() => toolbox.refresh());
```
Will
```suggestion
                        .then(toolbox.refresh);
```
work here?

kubama replied:
Looks like it works - changed.

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,43 @@
+    permission: Stage.GenericConfig.WIDGET_PERMISSION('labels'),
+    categories: [Stage.GenericConfig.CATEGORY.DEPLOYMENTS],
+
+    fetchParams() {
```
Yes, it's necessary

Gelio replied:
Could you tell me why :slightly_smiling_face: ? I did not dive deep, so an answer from you would save me some time researching

kubama replied:
I was convinced it was necessary (thus my initial response), but after double-checking it turned out it is not. Removed.

qooban commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                            content="Export"
+                            icon="external share"
+                            labelPosition="left"
+                            onClick={() =>
+                                Stage.Utils.saveAs(
+                                    new Blob([JSON.stringify(data.items.map(item => _.pick(item, 'key', 'value')))]),
+                                    `${data.deploymentId}-Labels.json`,
+                                    true
+                                )
+                            }
```
I would probably extract that to a function. What do you think?

kubama replied:
Could be - done.

qooban commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+export default function LabelsTable({ data, toolbox }) {
```
For new widgets/components let's maybe start putting strings to `en.json`. What do you think?
We'll have less work in the future.

kubama replied:
That's something new - please see https://github.com/cloudify-cosmo/cloudify-stage/pull/1208#discussion_r582676292.

qooban replied:
Yep, OK, the same answer as in https://github.com/cloudify-cosmo/cloudify-stage/pull/1208#discussion_r582676292 .

kubama commented on app/utils/stageUtils.js:
```diff
@@ -29,7 +29,7 @@ export default class StageUtils {
     static parseMarkdown = marked;
 
     static saveAs(...args) {
```
Addressed

qooban reviewed (commented):
I did partial review, didn't manage to check the behavior in the UI. Will try to check it tomorrow.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                                            value={currentLabelValue}
+                                            defaultValue={item.value}
+                                            popupContent="Revert to initial value"
+                                            onMouseDown={e => {
```
I'm using blur event to submit edited value (inline edit). I don't want the event to be triggered/handled on the revert icon click. Unfortunately, the click event is fired after blur, so it would be to late to do anything, as the `onBlur` handler would already run. The mousedown event however is fired before blur, so it's possible to prevent blur from being triggered by calling `preventDefault`.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                    <div className="content">
+                        Are you sure you want to remove label &nbsp;
+                        <Label size="large">
+                            {labelToDelete?.key} <span style={{ fontWeight: 'lighter' }}>{labelToDelete?.value}</span>
```
It is `undefined` when not set

Gelio replied:
Oh ok, because this expression is evaluated during the render, and not inside a callback :slightly_smiling_face: I did not notice it, thanks

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,43 @@
+
+        const formattedData = {
+            items: data,
+            deploymentId: toolbox.getContext().getValue('deploymentId')
+        };
+
+        return <LabelsTable data={formattedData} toolbox={toolbox} />;
```
Yes, it is handled - see LabelsTable internals

qooban replied:
OK, then please check if it works in all cases. I shared a video in the general comment: https://github.com/cloudify-cosmo/cloudify-stage/pull/1208#issuecomment-784862817 . 

I think that it would also be better to say that the deployment was not selected instead of "There are no Labels defined", which can be seen in two cases right now:
1. No deploymentId in the context
2. No labels attached to deployment
That will be in line with what we already do in other widgets - e.g. Blueprint Info or Deployment Info.

kubama replied:
Added specific message for no deployment selected case, added proper handling of deployment context switching.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                        </DataTable.Data>
+                        {hasManagePermission && (
+                            <DataTable.Data>
+                                <Icon
```
I'm not sure about this one, is replacing one icon with another a step towards usability? The 'save' icon would be followed by 'delete' icon, which is not related to the inline edit action. Do you have any thoughts on this @qooban?

Gelio replied:
Good point. The _Delete_ could also be removed, since it does not apply the user's "context" (being in an _inline edit_ mode). Thus, instead of the Confirm + Delete actions, we could show Save + Discard actions.

IMO that makes the UI adapt to the action the user changes, and I would find it more expected

qooban replied:
Hmm... Yeah, after doing some quick research I'm leaning towards @Gelio suggestion. 
For an inspiration for inline editing in general: 
https://www.patternfly.org/v4/components/inline-edit/design-guidelines/ - in particular this part: https://www.patternfly.org/v4/components/inline-edit/design-guidelines/#table-inline-edit
https://uxdesign.cc/designing-tables-for-reusability-490a3760533
Maybe we could have something like that?

But let's validate if any change here will not consume too much time. Inline editing is not critical feature in this widget. And we all know the tight deadlines for 6.0 content.

kubama replied:
The change was simple so it's implemented. I think it works and looks really well:
![image](https://user-images.githubusercontent.com/49753069/108983208-ea1af180-768e-11eb-9a59-6a968b6e4aab.png)

One note is that I removed automatic change submission on input blur - since we now have button for explicit change submission I found it unnecessary and even counterintuitive.

qooban replied:
Looks very good IMO 👍🏻

Gelio replied:
Yup, thanks for that :smile: Looks good

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                <DataTable.Column width="50%" label="Value" />
+                {hasManagePermission && <DataTable.Column width="80px" />}
+
+                {_.map(data.items, item => (
```
We don't - changed

qooban commented:
@kubama I did some tests in the UI. First of all thanks for the whole implementation. I spotted some areas which probably can be improved. In some cases I need to figure out the details with Ofer/Yoni.

1. **Inline editing** - it's very cool feature (probably we could have something similar in Secrets widget in the future). Few comments to it:
  a. We should have the same validation implemented as it is in LabelsInput for key and value inputs - input should be lowercased and only limited set of characters allowed. See `onSearchChange` in `common/src/labels/CommonDropdown.jsx` for details.
  b. It would be good to indicate that value is being changed (request sent to the manager) using widely used spinner icon somewhere in the widget (could be `toolbox.loading` or loading indicator inside the inline edit input field)
  c. We should provide error handling for adding input by showing the error in the top of the widget (or other way, but not hiding it in the console only). The common way we use is to have `ErrorMessage` above the `DataTable`, but you can develop something more contextual as well. 
Video about lack of error handling: 
https://user-images.githubusercontent.com/5202105/108963551-8b4b7d00-767a-11eb-9b70-4b6ae9049853.mp4

2. **Contextual deployment ID** - I think setting/switching deployment ID is not working fine.
See: https://user-images.githubusercontent.com/5202105/108963693-bafa8500-767a-11eb-979b-24c862377a8b.mp4

3. **Selectable table rows** - Is there any reason to use `cursor: pointer;` for the rows which are not going (or am I missing something) to be selected?

4. **Pagination, sorting, search** - good that you pointed out that pagination is not currently supported. I will take that with Yoni and figure out if that would be possible. I think it would be important to have it (as I remember from discussions with Ofer that we can have a lot of labels assigned). Also we should be able to sort the list by key alphabetically, so I'll check that with Yoni/Ofer as well. I think also it'd be good to have an option to search for a key using search box. What do you think? As we can have multiple values for the same key that could be quite useful, I think.

5. **Table rows alignment** - I found today that in the CLI they present values for the same key in the following way:
  ```
  Deployment labels
+------+---------------+
| key  |     values    |
+------+---------------+
| test |      test     |
| abc  | abaabab,ababa |
+------+---------------+
  ```
I know it was not part of the original design, I was not aware of such presentation until today, so for now I'd only like to ask you - do you think it would be very complex to implement it in similar way to CLI?
I'll take the potential requirement update with Ofer.

6. **Labels export format** - it's fine as you implemented it according to the design, but I'm wondering what's actually the use case for the Export. When I checked the CLI it expects the following:
```
Usage: cfy deployment labels add [OPTIONS] LABELS_LIST DEPLOYMENT_ID

  LABELS_LIST: <key>:<value>,<key>:<value>
```
so a string: `"<key>:<value>,<key>:<value>"`.
When I checked the API it expects the following:
```
PATCH http://<IP>/api/v3.1/deployments/AWS-Basics-VM-Setup; body: {"labels": [{"abc": "abaabab"}, {"abc": "ababa"}, {"test": "test"}, {"new": "new"}]}
```
so a list of objects: `[{<key>:<value>}, {<key>:<value>}]`.
I will take that with Ofer to clarify.

qooban commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -38,13 +40,17 @@ function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
         <Modal open={open} onClose={onHide}>
             <Modal.Header>
-                <Icon name="tags" /> {i18n.t('widgets.common.labels.modalHeader', { deploymentId })}
+                <Icon name="tags" /> {header || i18n.t('widgets.common.labels.modalHeader', { deploymentId })}
```
1. Maybe we can have two headers in `en.json` - e.g. one starting from 'Manage', the second one starting from 'Add' and use only `existingLabels` prop to determine which one to choose?
2. Maybe it would be also good to change the button name for add-mode to be 'Add' instead of 'Apply'?

Gelio replied:
Could we do

```suggestion
                <Icon name="tags" /> {header ?? i18n.t('widgets.common.labels.modalHeader', { deploymentId })}
```

to allow setting `header` to an empty string?

kubama replied:
> Could we do
> ...
> to allow setting `header` to an empty string?

I don't think an empty header would be a valid case.

kubama replied:
> 1. Maybe we can have two headers in `en.json` - e.g. one starting from 'Manage', the second one starting from 'Add' and use only `existingLabels` prop to determine which one to choose?

We could, but the effect of setting `existingLabels` would be a little bit surprising then.
Regarding strings organization, I would leave it as it is for now and refine it in the future - If I understand correctly this modal is going to be used in at least one more place. Once it is used everywhere where it's intended to it will be clear which strings are indeed common, and which are specific to particular widgets. It will be easier to accurately organize the strings then.

> 2. Maybe it would be also good to change the button name for add-mode to be 'Add' instead of 'Apply'?

Done.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+    function updateLabelValue() {
+        if (!currentLabelValue) return;
+
+        if (_.find(data.items, { value: currentLabelValue })) {
```
> Does it mean there cannot be 2 labels with different keys but the same value?

Actually you spotted a bug - fixed.

> Also, could we import the find from lodash instead of accessing a global variable 🙂 ? What's your take on this approach?

With current implementation 3 different lodash functions are used in this file, so I'd stay with referencing the global rather then importing each of the functions.

Gelio replied:
> With current implementation 3 different lodash functions are used in this file, so I'd stay with referencing the global rather then importing each of the functions.

Alright :slightly_smiling_face: You can consider using `import _ from 'lodash';` then, just FYI. I encourage using ESM imports and not global variables, since they are more explicit and more inline with what we do in other parts of the codebase and with other imports :slightly_smiling_face:

kubama replied:
OK, import added.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+                        <DataTable.Data>{item.key}</DataTable.Data>
+                        <DataTable.Data>
+                            {item === labelInEdit ? (
+                                <Form.Input
```
Done

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,161 @@
+
+    const hasManagePermission = Stage.Utils.isUserAuthorized('deployment_create', toolbox.getManagerState());
+
+    return (
```
Done

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,43 @@
+
+    fetchParams() {
+        return {};
+    },
```
No - removed.

Gelio commented:
I'll reread this PR after the standup :slightly_smiling_face:

Gelio commented:
@qooban 
>Selectable table rows - Is there any reason to use cursor: pointer; for the rows which are not going (or am I missing something) to be selected?

That's something I noticed in all of the tables :smile: 

The problem with `cursor: pointer` lies in `cloudify-ui-components`:
https://github.com/cloudify-cosmo/cloudify-ui-components/blob/98dd2d04db6051f12702ded5e378ceb53989e68c/src/components/data/DataTable/TableRow.jsx#L39

`onClick` has a default value set in `defaultProps`, so it is always a function.

kubama commented on widgets/common/src/LabelsInput.jsx:
```diff
@@ -102,12 +105,14 @@ export default function LabelsInput({ initialLabels, onChange, toolbox }) {
 }
 
 LabelsInput.propTypes = {
+    addMode: PropTypes.bool,
     initialLabels: LabelsPropType,
```
In `addMode` the `initialLabels` still contains labels that are already set - the difference is that they are not displayed, but still used for detecting duplicates.

Gelio replied:
I'm afraid that's not the case. If you have existing labels, then `initialLabels` will not be empty, but you are still in append (add-only) mode :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/889383/109020866-a76f0e80-76ba-11eb-85c5-87275d84cfbe.png)

kubama commented on widgets/common/src/LabelsInput.jsx:
```diff
@@ -64,15 +65,17 @@ export default function LabelsInput({ initialLabels, onChange, toolbox }) {
-                    style={{ ...iconStyle, right: '2em' }}
-                />
+                {!addMode && (
+                    <RevertToDefaultIcon
```
I'm not sure - in LabelsTable we have a context of editing an existing value, while here we have a context of adding multiple new labels, where the initial state is an empty list. So there is no 'initial' or 'default' state one can revert to.
To be honest, I don't think this icon makes much sense even in standard 'manage' mode - it would if there were other inputs in this modal. In such case user might want to revert changes in labels input while submitting changes made to other inputs. But since there is only labels input in this modal then I don't see a use case for the icon.

Gelio commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -1,4 +1,4 @@
-function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
+function ManageLabelsModal({ deploymentId, existingLabels, header, open, onHide, toolbox }) {
```
I'm wondering... Maybe instead of having a component that sometimes fetches its data, and sometimes does not, would it not be cleaner to have 2 components:
1. The existing `ManageLabelsModal` that receives `existingLabels` and does not fetch its data
2. A new component that fetches labels and displays `ManageLabelsModal`

This way in the new Labels component we could render the first one, and the other place that needs the data-fetching part could use the 2nd one.

IMO that's cleaner than doing different things in the component :smile:

kubama replied:
The situation is a bit specific here - it's the first time ManageLabelsModal is used in any widget and at the same time as you can see it was implemented in a way that makes it suitable to be used in a different context. I'm not sure what further changes will be needed once its further reused. I would say its implementation should be revisited and refactored once this is clear, not necessarily now.

Gelio replied:
I am not fully convinced here, because we are putting more logic into the component and forcing it to have 2 modes, meaning sometimes it's controlled, and sometimes uncontrolled. This makes it harder to understand and adjust in the future too.

I would really prefer if we had 2 different components, providing a hierarchy here, instead of forcing a single component to behave in 2 ways :slightly_smiling_face: 

This is not as strong as to block the PR, but I encourage you to rethink it, and if you decide you stick with the current way - so be it :slightly_smiling_face:

kubama replied:
I understand and agree - I'm just saying it would be better if this component was refactored by someone who implemented usage in management mode - since it's not there yet it's hard to say what he exact direction should be. So please consider the current state as temporary.

Gelio replied:
Ok sure :slightly_smiling_face:

Gelio commented on widgets/common/src/LabelsInput.jsx:
```diff
@@ -102,12 +105,14 @@ export default function LabelsInput({ initialLabels, onChange, toolbox }) {
 };
 
 LabelsInput.defaultProps = {
+    addMode: false,
```
The functionality sounds more like `appendMode` :smile:

Gelio replied:
Or, to make it more descriptive, we could do

```suggestion
    appendToExistingMode: false,
```

kubama replied:
In the context of the widget - it's an append mode. But in the context of this particular component I wouldn't say so - appending means adding something to existing items, while in `addMode` behaviour of this components is actually shifted in an opposite direction - labels passed in `initialLabels` are actually being hidden from the view.

Gelio replied:
Ok, I see your point. To me, _add mode_ is still cryptic, because we are adding labels anyway :slightly_smiling_face: Maybe we could settle on renaming it to something else, perhaps `hideInitialLabels`?

kubama replied:
Sounds good - changed.

Gelio reviewed (commented):
Thanks for addressing all the previous comments :slightly_smiling_face: 

This time I've put some more time into the review, hence, more comments :slightly_smiling_face:

kubama commented on widgets/common/src/RevertToDefaultIcon.jsx:
```diff
@@ -29,6 +30,16 @@ RevertToDefaultIcon.propTypes = {
      */
     onClick: PropTypes.func,
 
+    /**
+     * function to be called on revert icon mouse down
+     */
+    onMouseDown: PropTypes.func,
```
I don't want the inline edit input to lost focus on revert icon click. Unfortunately, the click event is fired after blur. The mousedown event however is fired before blur, so it's possible to prevent blur from being triggered by calling preventDefault.

qooban commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,51 @@
+    id: 'labels',
+    name: 'Labels',
+    description: 'This widget shows a list of defined labels',
+    initialWidth: 12,
+    initialHeight: 24,
```
`initialWidth: 12` means basically "take whole width of the container". We use that in many widgets - I'm OK with that, it make sense for me.

NOTE: Resolution of width and height is different, so width = 12 and height = 12 will not make a square.

Gelio replied:
I was pointing out the fact that the content of the table is pretty small :slightly_smiling_face: There are only 2 columns, with 2 strings, most likely short ones, meaning that there is a vast amount of whitespace, at least when I was testing it on a full HD screen with the Labels widget being the only widget in the row

![image](https://user-images.githubusercontent.com/889383/109124045-c61ce600-774a-11eb-899d-d4139c99a780.png)

Looks worse to me than if it would start at half the width as a sensible defautl :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/889383/109124156-e64ca500-774a-11eb-9acb-671c478b703d.png)

qooban replied:
Minor thing. User can always change the default value. Let Kuba M. decide what do with it.

kubama replied:
These initial dimensions are used only when widget is manually added to a page. At the moment all table-based widgets have same initial dimensions. I think it is a good default - if it was narrower the page layout would have empty spaces just after adding widget(s), which wouldn't look that great for a default.
When widget is added to a predefined page it has its dimensions set anyway, so it's a different story.

Gelio replied:
Yup, you convinced me :slightly_smiling_face: Thanks

Gelio commented:
Some bugs/problems I have found when trying out extreme inputs:

1. The error message is invalid when passing a very long label key (the string was `verylongkey` repeated multiple times, and the error message said something about **the value** containing illegal characters (it was `hey` or `abc`))

    ![image](https://user-images.githubusercontent.com/889383/109124449-388dc600-774b-11eb-9ec7-fccd0e9a7367.png)

2. The component does display long values very well:

    ![image](https://user-images.githubusercontent.com/889383/109124677-78ed4400-774b-11eb-8bca-f0ef96ac3011.png)

    This happens in the rows themselves, but also in the modal
    
    ![image](https://user-images.githubusercontent.com/889383/109124720-886c8d00-774b-11eb-8349-8a3d3060be78.png)
    
    ![image](https://user-images.githubusercontent.com/889383/109124762-97533f80-774b-11eb-9cdc-8fb6d12ee2db.png)
    
    ![image](https://user-images.githubusercontent.com/889383/109124792-a0dca780-774b-11eb-88dc-bd56d5c18d44.png)

kubama commented:
>    a. We should have the same validation implemented as it is in LabelsInput for key and value inputs - input should be lowercased and only limited set of characters allowed. See `onSearchChange` in `common/src/labels/CommonDropdown.jsx` for details.

Done.

>    b. It would be good to indicate that value is being changed (request sent to the manager) using widely used spinner icon somewhere in the widget (could be `toolbox.loading` or loading indicator inside the inline edit input field)

Done.

>    c. We should provide error handling for adding input by showing the error in the top of the widget (or other way, but not hiding it in the console only). The common way we use is to have `ErrorMessage` above the `DataTable`, but you can develop something more contextual as well.

After addressing point a. I think there is no use case for showing an error message. 

> 2. **Contextual deployment ID** - I think setting/switching deployment ID is not working fine.

Fixed.

> 3. **Selectable table rows** - Is there any reason to use `cursor: pointer;` for the rows which are not going (or am I missing something) to be selected?

No reason, changed.

> 4. **Pagination, sorting, search** - good that you pointed out that pagination is not currently supported. I will take that with Yoni and figure out if that would be possible. I think it would be important to have it (as I remember from discussions with Ofer that we can have a lot of labels assigned). Also we should be able to sort the list by key alphabetically, so I'll check that with Yoni/Ofer as well.

Sorting is now implemented.

> I think also it'd be good to have an option to search for a key using search box. What do you think? As we can have multiple values for the same key that could be quite useful, I think.

I think search box is mainly usable with pagination. Also, with its default implementation it uses server-side search, which again is not supported by labels API.

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,51 @@
+
+    fetchParams(widget, toolbox) {
+        return {
+            deploymentId: toolbox.getContext().getValue('deploymentId')
```
The benefit is data refetch on deploymentId change.

Gelio replied:
Clever :smile:

Gelio replied:
I'm not sure if it's worth it, but maybe we could document it?

When seeing such code, I would be tempted to move it to the `fetchData` function and break this functionality

kubama replied:
Why not, added the comment.

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,51 @@
+            return <Loading />;
+        }
+
+        const deploymentId = toolbox.getContext().getValue('deploymentId');
+        if (!deploymentId) {
+            const { Message } = Stage.Basic;
+            return <Message info>No deployment selected</Message>;
+        }
```
Done.

kubama commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        const formattedData = {
+            items: data.map(item => _.pick(item, 'key', 'value')),
```
Done.

kubama commented on widgets/labels/widget.css:
```diff
@@ -0,0 +1,7 @@
+.labelsWidget .labelValueEditInput input {
+    padding: 4px;
```
This value is set in relation to font size, `input` border and `td` padding. Since border is set in `px` anyway the proposed change doesn't bring any value, it just makes things cryptic.

Gelio replied:
Sure, sounds reasonable :slightly_smiling_face:

kubama commented on widgets/labels/src/LabelValueInput.jsx:
```diff
@@ -0,0 +1,44 @@
+            autoFocus
+            focus
+            fluid
+            style={{ padding: 0, marginLeft: -5, marginRight: -5 }}
```
The `className` here is not to set styling for this particular component but to be able to reference its inner `input` in `widget.css`.

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,157 @@
+
+    const actions = new DeploymentActions(toolbox);
+
+    useEffect(() => setLabels(data.items), [JSON.stringify(data.items)]);
```
I believe it's a commonly recognized and used pattern. It even has an 'official' recommendation:
https://twitter.com/dan_abramov/status/1104414469629898754?lang=en

Gelio replied:
I'm not saying that it's wrong :slightly_smiling_face: My point here was about the possible performance implications :slightly_smiling_face: If we have to serialize the labels upon every rerender (and the table looks to be rerendering upon every keystroke when editing), that can introduce jank, and unnecessary operations

What I'm thinking we could do to mitigate it is:
1. In `widget.jsx`, memoize `labels`, so that the reference to the array changes only if there's new data coming from the network (we can use `isEqual` or `JSON.stringify` there)
2. Here, use `data.items` directly, without having to serialize on every rerender

Now that I think about it, I don't see why we need put `labels` in the state of `LabelsTable` :thinking: I only briefly took a look, but I did not see why

kubama replied:
> What I'm thinking we could do to mitigate it is:
> 
> 1. In `widget.jsx`, memoize `labels`, so that the reference to the array changes only if there's new data coming from the network (we can use `isEqual` or `JSON.stringify` there)
> 2. Here, use `data.items` directly, without having to serialize on every rerender

I'm not convinced. Making the change suggested in point 2. would make this component work incorrectly when used in regular way (without memoization). As components should be self-contained units this would be a flaky design.

> Now that I think about it, I don't see why we need put `labels` in the state of `LabelsTable` 🤔 I only briefly took a look, but I did not see why

Have a look at `updateLabelValue` - for a more responsive UI labels are updated before backend update request is completed.

Gelio replied:
> I'm not convinced. Making the change suggested in point 2. would make this component work incorrectly when used in regular way (without memoization). As components should be self-contained units this would be a flaky design.

Ok, we can revisit the memorization part if there will be performance issues :slightly_smiling_face:

kubama commented on widgets/labels/src/LabelValueInput.jsx:
```diff
@@ -0,0 +1,44 @@
+    const { RevertToDefaultIcon } = Stage.Common;
+    const { useInput } = Stage.Hooks;
+
+    const [currentLabelValue, setCurrentLabelValue] = useInput(initialValue);
```
I believe this comment is no longer valid after adding 'submit' button in inline edit mode.

Gelio replied:
Well, yes and no :smile: We still have duplicated state, but I don't have a good way of fixing that :slightly_smiling_face: Let's leave it be

kubama commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,157 @@
+                    <DataTable.Row key={`${item.key}:${item.value}`}>
+                        <DataTable.Data>{item.key}</DataTable.Data>
+                        <DataTable.Data>
+                            {_.isEqual(item, labelInEdit) ? (
```
Equality check here is closer to what we actually want to check here - it's not important if the two object references are identical - it's important if objects are equal. So even if `===` did the job here using it would mean working with an implicit assumption.
Also, please note that `_.isEqual` checks for references equality before performing actual equality check. But even if objects are not _same_ in this case they contain only two properties, so I wouldn't call eventual comparison expensive.

Gelio replied:
I did not know `isEqual` does a reference check first :smile: thanks for sharing, makes sense :slightly_smiling_face: 

Anyway, I was concerned about performance, but possibly prematurely :slightly_smiling_face: Let's leave it

kubama commented:
> Some bugs/problems I have found when trying out extreme inputs:
> 
> 1. The error message is invalid when passing a very long label key (the string was `verylongkey` repeated multiple times, and the error message said something about **the value** containing illegal characters (it was `hey` or `abc`))

I added length limitation to the `useLabelInput` hook - should be no longer an issue.

> 2. The component does display long values very well:

Added overflow handling to table cells.

>    This happens in the rows themselves, but also in the modal

This seems to be a general issue with dropdown component, I believe fixing this is out of scope of this PR. Could you please create a separate JIRA ticket to address this issue?

Gelio commented:
Sorry, due to unexpected issues with #1210 I didn't have enough time today to review this PR :worried: I'll take a look at it on Monday

Gelio commented on widgets/labels/src/widget.jsx:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        const formattedData = {
+            labels: _(data)
+                .map(item => _.pick(item, 'key', 'value'))
+                .sortBy('key')
+                .value(),
```
Clean use of chaining :+1:

Gelio commented on widgets/labels/src/LabelsTable.jsx:
```diff
@@ -0,0 +1,144 @@
+                {hasManagePermission && <DataTable.Column width="80px" />}
+
+                {labels.map(item => (
+                    <DataTable.Row key={`${item.key}:${item.value}`} onClick={null}>
```
I believe you no longer need the artificial `onClick` :slightly_smiling_face: It was fixed in ui-components

```suggestion
                    <DataTable.Row key={`${item.key}:${item.value}`}>
```

kubama replied:
Right, removed.

Gelio reviewed (commented):
Some more comments on the latest changes :slightly_smiling_face:

Gelio commented:
> This seems to be a general issue with dropdown component, I believe fixing this is out of scope of this PR. Could you please create a separate JIRA ticket to address this issue?

Sure :slightly_smiling_face:  https://cloudifysource.atlassian.net/browse/RD-1575

kubama commented on widgets/common/src/hooks/useLabelInput.js:
```diff
@@ -0,0 +1,30 @@
+function useLabelInput(onChange, initialValue = '') {
+    const { useBoolean, useResettableState } = Stage.Hooks;
+    const [inputValue, setInputValue, resetInputValue] = useResettableState(initialValue);
+    const [isInvalidCharacterTyped, setInvalidCharacterTyped, unsetInvalidCharacterTyped] = useBoolean();
```
Done

Gelio commented on widgets/common/src/RevertToDefaultIcon.jsx:
```diff
@@ -39,6 +50,8 @@ RevertToDefaultIcon.defaultProps = {
     value: undefined, // value can be undefined
     defaultValue: undefined, // defaultValue can be undefined
     onClick: _.noop,
+    onMouseDown: _.noop,
```
One thing I want to bring up generally is that I often see noops assigned as default functions. I agree with it when the function is actually required, but, to me, it does not make sense when the function is optional, such as here :slightly_smiling_face: 

If `Icon` can happily accept `undefined` instead of a valid function (I have not checked that in the code, but I am pretty sure it will), then we can save ourselves attaching event listeners (even if they are quite cheap) and the small overhead of calling them.

Thus, I would encourage using `undefined` as a default value for optional callbacks, rather than a valid function that does nothing :slightly_smiling_face: 

Assigning default noop functions was what caused a bug to appear in ui-components that was later fixed by https://github.com/cloudify-cosmo/cloudify-ui-components/pull/59

```suggestion
    onClick: undefined,
    onMouseDown: undefined,
```

kubama replied:
Yes, I agree. Changed.

kubama commented on widgets/common/src/hooks/useLabelInput.js:
```diff
@@ -0,0 +1,30 @@
+        isInvalidCharacterTyped,
+        submitChange: (event, data) => {
+            // supports both dropdown as well as regular input
+            const lowercasedNewTypedValue = _.toLower(data.searchQuery ?? data.value).substr(0, 256);
```
It's used in one place and it's pretty obvious from the context what is the meaning of the value, so I don't see much value in such change, but let it be - updated.

kubama commented on widgets/common/src/labels/ValidationErrorPopup.jsx:
```diff
@@ -19,3 +19,8 @@ export default function ValidationErrorPopup({ open }) {
 };
+
+Stage.defineCommon({
+    name: 'LabelValidationErrorPopup',
```
Done.

kubama commented on widgets/common/src/labels/ValidationErrorPopup.jsx:
```diff
@@ -1,4 +1,4 @@
-export default function ValidationErrorPopup({ open }) {
```
Defining a common is already a way of exporting, so I guess it's less confusing to use a single way. Also for consistency - we have multiple common files that are also used directly inside `common`, and we always reference them through `Stage`.

Gelio replied:
Ok, right, thanks for providing that info :slightly_smiling_face:

Gelio commented:
Looks good to me right now :slightly_smiling_face: 

@kubama are you planning on addressing the TODO comments in the PR description in this PR?

kubama commented:
> @kubama are you planning on addressing the TODO comments in the PR description in this PR?

I do 🙂.

Gelio commented:
@kubama Great :slightly_smiling_face: I'm thinking - maybe we split the task into 2 PRs? This way at least this part will be in master and will be gone from the diff :slightly_smiling_face: That would certainly be helpful when it comes to reviewing the PR again, since there are more changes planned. What's your take on this?

If you do not want to expose a half-baked Labels widget yet, we can do that behind a flag, similar to `DeploymentsViewWidget`:

https://github.com/cloudify-cosmo/cloudify-stage/blob/cc53e7830a7950025afe016ec9534c32fba9e9e9/widgets/deploymentsView/src/widget.tsx#L28-L123

kubama commented:
@Gelio The widget itself is completed, it just lacks documentation, Cypress test and `csys` labels handling - I think it can be merged indeed without the need for hiding it.
@qooban If you also agree then please approve and we are good to go.

qooban commented on widgets/common/src/ManageLabelsModal.jsx:
```diff
@@ -1,4 +1,4 @@
-function ManageLabelsModal({ deploymentId, open, onHide, toolbox }) {
+function ManageLabelsModal({ deploymentId, existingLabels, header, applyButtonContent, open, onHide, toolbox }) {
```
We can revisit the set of props in the future. I think it may be possible to simplify it a bit.

qooban commented on widgets/common/src/hooks/useLabelInput.js:
```diff
@@ -0,0 +1,30 @@
+const allowedCharacters = /^[a-z][a-z0-9._-]*$/i;
+
+function useLabelInput(onChange, initialValue = '') {
```
Great idea, I like it, too 👍🏻

qooban commented on widgets/common/src/labels/LabelValidationErrorPopup.jsx:
```diff
@@ -16,6 +16,11 @@ export default function ValidationErrorPopup({ open }) {
 }
 
-ValidationErrorPopup.propTypes = {
+LabelValidationErrorPopup.propTypes = {
```
I see what was the reason behind it.
Maybe in the future we can have a `Labels` object provided by `Stage.Common` with all labels related components inside.

qooban commented on widgets/labels/src/DeleteConfirmModal.jsx:
```diff
@@ -0,0 +1,41 @@
+            }}
+            content={
+                <div className="content">
+                    Are you sure you want to remove label &nbsp;
```
I assume that this (and all other label strings for new widget) can be migrated to `en.json` within the next PR.

qooban reviewed (approved):
Approved 👍🏻 

I didn't go very deep in the review. I agree it's getting harder to get with it while it's growing. 
Added some comments which can be addressed in the later phase (next PR or even later).

kubama commented:
> Added some comments which can be addressed in the later phase (next PR or even later).

Ok, I will address the comments in the next PR.
